File: bindings/wasm/benchmarks/bench_lookup.mjs
import { bench, group, run } from "mitata";
import { country_code } from "../pkg-nodejs/iptocc_wasm.js";

const V4 = [
  ["afrinic",      "41.0.0.1"],
  ["apnic",        "1.0.16.1"],
  ["arin",         "8.8.8.8"],
  ["lacnic",       "200.160.0.1"],
  ["ripencc",      "193.0.6.139"],
  ["miss_private", "10.0.0.0"],
];

const V6 = [
  ["afrinic",       "2001:4200::1"],
  ["apnic",         "2001:200::1"],
  ["arin",          "2001:4860:4860::8888"],
  ["lacnic",        "2001:1280::1"],
  ["ripencc",       "2001:67c:18::1"],
  ["miss_loopback", "::1"],
];

for (let i = 0; i < 200000; i++) {
  for (const [, addr] of V4) country_code(addr);
  for (const [, addr] of V6) country_code(addr);
}

group("v4", () => {
  for (const [name, addr] of V4) {
    bench(name, () => country_code(addr));
  }
});

group("v6", () => {
  for (const [name, addr] of V6) {
    bench(name, () => country_code(addr));
  }
});

const BATCH_SIZES = [10, 100, 1000, 10000];

const repeat = (cases, n) => {
  const addrs = [];
  while (addrs.length < n) {
    for (const [, addr] of cases) addrs.push(addr);
  }
  return addrs.slice(0, n);
};

group("batch_v4", () => {
  for (const n of BATCH_SIZES) {
    const addrs = repeat(V4, n);
    bench(`n=${n}`, () => country_code(addrs));
  }
});

group("batch_v6", () => {
  for (const n of BATCH_SIZES) {
    const addrs = repeat(V6, n);
    bench(`n=${n}`, () => country_code(addrs));
  }
});

await run();

File: bindings/python/benchmarks/bench_lookup.py
import pytest

V4_CASES = [
    ("afrinic", "41.0.0.1", "ZA"),
    ("apnic", "1.0.16.1", "JP"),
    ("arin", "8.8.8.8", "US"),
    ("lacnic", "200.160.0.1", "BR"),
    ("ripencc", "193.0.6.139", "NL"),
    ("miss_private", "10.0.0.0", None),
]

V6_CASES = [
    ("afrinic", "2001:4200::1", "ZA"),
    ("apnic", "2001:200::1", "JP"),
    ("arin", "2001:4860:4860::8888", "US"),
    ("lacnic", "2001:1280::1", "BR"),
    ("ripencc", "2001:67c:18::1", "NL"),
    ("miss_loopback", "::1", None),
]

BATCH_SIZES = [10, 100, 1000, 10000]


@pytest.fixture(scope="session")
def ours():
    import iptocc

    return iptocc.country_code


@pytest.mark.parametrize(("name", "addr", "expected"), V4_CASES, ids=[c[0] for c in V4_CASES])
def test_iptocc_v4(benchmark, ours, name, addr, expected):
    benchmark.group = "v4_single"
    result = benchmark(ours, addr)
    assert result == expected


@pytest.mark.parametrize(("name", "addr", "expected"), V6_CASES, ids=[c[0] for c in V6_CASES])
def test_iptocc_v6(benchmark, ours, name, addr, expected):
    benchmark.group = "v6_single"
    result = benchmark(ours, addr)
    assert result == expected


@pytest.mark.parametrize("size", BATCH_SIZES, ids=[f"n={n}" for n in BATCH_SIZES])
def test_iptocc_batch(benchmark, ours, size):
    benchmark.group = "batch"
    addrs = [c[1] for c in V4_CASES] * (size // len(V4_CASES) + 1)
    addrs = addrs[:size]
    result = benchmark(ours, addrs)
    assert len(result) == size
